packages/core/lib/messages/MetadataFiller.ts

```diff
@@ -0,0 +1,62 @@
+import { randomUUID } from 'node:crypto'
+
+import type { BaseEventType } from '../events/baseEventSchemas'
+import type { CommonEventDefinition } from '../events/eventTypes'
+
+import type { MessageMetadataType } from './baseMessageSchemas'
+
+export type IdGenerator = () => string
+export type TimestampGenerator = () => string
+
+export type MetadataFillerOptions = {
+  serviceId: string
+  schemaVersion: string
+  idGenerator?: IdGenerator
+  timestampGenerator?: TimestampGenerator
+  defaultVersion?: string
+}
+
+export type MetadataFiller<
+  T extends BaseEventType = BaseEventType,
+  D = CommonEventDefinition,
+  M = MessageMetadataType,
+> = {
+  produceMetadata(currentMessage: T, eventDefinition: D, precedingMessageMetadata?: M): M
+  produceId(): string
+  produceTimestamp(): string
+}
+
+export class CommonMetadataFiller implements MetadataFiller {
+  private readonly serviceId: string
+  public readonly produceId: IdGenerator
+  public readonly produceTimestamp: TimestampGenerator
+  private readonly defaultVersion: string
+
+  constructor(options: MetadataFillerOptions) {
+    this.serviceId = options.serviceId
+    this.defaultVersion = options.defaultVersion ?? '1.0.0'
```

Comment:
Do we want to migrate to SerVer for the `schemaVersion` field? to me it sounds like the best approach but asking just to confirm

Reply:
ondering if could be beneficial, on consumer, to differentiate between `schemaVersion` to use one handler or the other 🤔 (definitely not part of this PR, sorry for raising things are not relevant here)

Reply:
hmmm, interesting idea! and yes, that would be awesome for the compatibility. will implement in a separate PR